### PR TITLE
Add support for PolyLoss (ICLR 2022)

### LINKED ISSUE #466: Add support for PolyLoss (ICLR 2022)
Hi, thanks for maintaining the library!
What are your thoughts on adding support for PolyLoss (see https://arxiv.org/pdf/2204.12511.pdf, page 10)?

Though it can be implemented externally based on the existing `softmax_cross_entropy`, I would still suggest to add it to the library alongside with some tests

I would also like to volunteer to work on that, if the change seems reasonable :)

### PR DESCRIPTION
**Pull request by acforvs**
Closes #466 

I focused on implementing the Poly-1 cross entropy loss with alpha-label smoothing since, as it is shown in the paper, the first term plays the most important role. Also, I crafted a small test showing that the behavior is indeed similar to the one of the `softmax_cross_entropy` when `\epsilon = 0`

I didn't add it to `optax/__init__.py` and `docs/api.rst` files yet (would it be ok to do that in a separate PR?)

Please, let me know what you think!

### CODE REVIEW

File: optax/_src/loss_test.py [hunk 1]
```diff
@@ -496,5 +496,85 @@ def test_batched(self):
...
+    self.labels = np.array([0.1, 0.15, 0.2, 0.25, 0.3])
+
+    self.batched_logits = np.array([[4.0, 2.0, 1.0], [0.0, 5.0, 1.0]])
+    self.batched_labels = np.array([[1.0, 0.0, 0.0], [0.0, 0.8, 0.2]])
+    # all expected values are computed using tf version of `poly1_cross_entropy`
+    # see page 10 here https://arxiv.org/pdf/2204.12511.pdf for more
+
+  @chex.all_variants
+  @parameterized.parameters(
+      dict(eps=2, alpha=0, expected=4.531657285679147),
```
Comment (mkunesch) [inline optax/_src/loss_test.py:514]: Since we are testing to atol 1e-4 we can remove some of the figures from the expected results.

**Review by mkunesch — COMMENTED** [review]
Thanks a lot for this contribution - this looks great! Thanks especially for adding the maths and recommended parameters.

I've added a few comments. I'll do another pass to compare the maths to the paper later.

File: optax/_src/loss.py [hunk 2]
```diff
@@ -519,3 +519,58 @@ def hinge_loss(predictor_outputs: chex.Array,
...
+      alpha: The smoothing factor, the greedy category with be assigned
+        probability `(1-alpha) + alpha / num_categories` (default = 0.0)
+
+    Returns:
+      poly loss between each prediction and the corresponding target
+      distributions, with shape `[...]`.
+    """
+  chex.assert_type([logits], float)
+
+  smoothed_labels = smooth_labels(labels=labels, alpha=alpha)
```
[the remark this answers is not in the document]
Comment (acforvs) [reply optax/_src/loss.py]: I agree, this is indeed an independent functionality; I removed it and fixed the tests accordingly

File: optax/_src/loss.py [hunk 3]
```diff
@@ -519,3 +519,58 @@ def hinge_loss(predictor_outputs: chex.Array,
...
+        one hot encoding specifying the correct class for each input;
+        must have a shape broadcastable to `[..., num_classes]``
+      epsilon: The coefficient of the first polynomial term (default = 2.0).
+        According to the paper, the following values are recommended:
+        - For the ImageNet 2d image classification, epsilon = 2.0
+        - For the 2d Instance Segmentation and object detection, epsilon = -1.0
+        - It is also recommended to adjust this value
+          based on the task and dataset at hand. For example, one can use
+          simple grid search to achieve it.
+      alpha: The smoothing factor, the greedy category with be assigned
```
[the remark this answers is not in the document]
Comment (acforvs) [reply optax/_src/loss.py]: My bad, I copypasted the description from [here](https://github.com/deepmind/optax/blob/master/optax/_src/loss.py#L107) without going through it 

Since I removed the smoothing itself, this issue is now fixed in this PR. Should I submit a separate PR to fix this line for the `smooth_labels` as well?

**Comment by acforvs** [comment]
Thanks for the review, I updated the code!

**Review by mkunesch — APPROVED** [review]
Hi! Sorry for dropping the ball on this! I've added a few small nits, but we we can also address those during merging so I've also approved.

**Comment by acforvs** [comment]
@mkunesch addressed the comments. The tests seem to be failing locally though, but it looks like the problem is in `ctc_loss_with_forward_probs`

I'll do another pass in a separate PR to fix the style in other docstrings if that's ok

**Comment by mkunesch** [comment]
Hi! Thanks a lot for the changes and sorry for the delay!

> The tests seem to be failing locally though, but it looks like the problem is in ctc_loss_with_forward_probs
Yes, that was an unrelated problem. We've fixed that error now - could you sync with master? That should make the checks pass so that we can merge. 

> I'll do another pass in a separate PR to fix the style in other docstrings if that's ok
Sounds good, thanks a lot!

**Review by mkunesch — APPROVED** [review]

**Comment by acforvs** [comment]
Hi, no worries!
I synced my local branch with master, but it didn't trigger the tests for some reason - could you re-run them please?

**Comment by acforvs** [comment]
Hi @mkunesch, thanks for the review! Should I also update https://github.com/deepmind/optax/blob/master/docs/api.rst#L494 and https://github.com/deepmind/optax/blob/master/optax/__init__.py#L96?

**Comment by mkunesch** [comment]
Ah, yes please - I forgot to check that in my review. Thanks a lot!

**Comment by acforvs** [comment]
Added here: https://github.com/deepmind/optax/pull/537
